File: chapters/integration/xmlrpc-client/xmlrpc-client-store/src/main/java/exist/book/example/xmlrpc/client/StoreApp.java
package exist.book.example.xmlrpc.client;

import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.net.URL;
import javax.xml.bind.DatatypeConverter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import redstone.xmlrpc.XmlRpcClient;

public class StoreApp {
    
    private final static Logger logger = LoggerFactory.getLogger(StoreApp.class);
    
    public static void main(final String[] args) throws Exception {
        
        //get the command line arguments
        if(args.length < 6) {
            printUseage();
            System.exit(1);
            return;
        }
        final String server = args[0];
        final int port = Integer.parseInt(args[1]);
        final File file = new File(args[2]);
        if(!file.exists()) {
           System.err.println("File '" + file.getAbsolutePath() + "' does not exist!");
           System.exit(2);
        }
        final String mediaType = args[3];
        final String collection = args[4];
        final String username = args[5];
        final String password;
        if(args.length == 7) {
            password = args[6];
        } else {
            password = "";
        }
        
        //get an XML-RPC connection to the eXist server
        final String uri = "http://" + server + ":" + port + "/exist/xmlrpc";
        final XmlRpcClient rpc = new XmlRpcClient(new URL(uri), true);
        
        //setup HTTP Basic Authentication
        final String auth = DatatypeConverter.printBase64Binary((username + ":" + password).getBytes());
        rpc.setRequestProperty("Authorization", "Basic " + auth);
        
        logger.info("Starting upload of {} to {}{}...", file.getAbsolutePath(), uri, collection);
        
        //upload the resource
        String remoteIdentifier = null;
        final byte buf[] = new byte[1024];
        int read = -1;
        InputStream is = null;
        try {
            is = new FileInputStream(file);            
            while((read = is.read(buf)) > -1) {
                if(remoteIdentifier == null) {
                    //upload first chunk of file to server
                    remoteIdentifier = (String)rpc.invoke("upload", new Object[] { buf, read });
                } else {
                    //append further chunks of file to remote file on server
                    remoteIdentifier = (String)rpc.invoke("upload", new Object[]{ remoteIdentifier, buf, read });
                }
            }
        } finally {
            if(is != null) {
                is.close();
            }
        }
        
        //have the server parse and store the remote file into the database
        final Boolean result = (Boolean)rpc.invoke("parseLocal", new Object[] { remoteIdentifier, collection + "/" + file.getName(), true, mediaType });
        if(result.booleanValue()) {
            logger.info("Finished upload OK.");     
        } else {
            logger.error("Could not store file.");
            System.exit(3);
        }
    }
    
    private static void printUseage() {
        System.err.println();
        System.err.println("useage: StoreApp <server> <port> <file> <internet-media-type> <collection> <username> [password]");
        System.err.println("\tserver: The hostname or ip address of the server e.g. localhost.");
        System.err.println("\tport: The tcp port of the server e.g. 8080.");
        System.err.println("\tfile: The path to a file you wish to store in the database.");
        System.err.println("\tinternet-media-type: The Internet Media Type of the file you wish to store in the database.");
        System.err.println("\tcollection: The path to a collection in the database where the file should be stored e.g. /db/mycollection.");
        System.err.println("\tusername: The eXist account to use to connect to the database e.g. admin.");
        System.err.println("\tpassword: The password of the eXist account to use to connect to the database, if ommitted then an blank password is used.");
        System.err.println();
    }
}

File: chapters/integration/xmlrpc-client/StoreApp.py
#!/usr/bin/python

##
# A direct port of the StoreApp Java XML-RPC Client example
# shipped with the eXist O'Riley book
#
# Author: Adam Retter
##

import array
import os
import sys
import xmlrpclib

def printUseage():
	sys.stderr.write("""useage: StoreApp.py <server> <port> <file> <internet-media-type> <collection> <username> [password]

	server: The hostname or ip address of the server e.g. localhost.
	port: The tcp port of the server e.g. 8080.
	file: The path to a file you wish to store in the database.
	internet-media-type: The Internet Media Type of the file you wish to store in the database.
	collection: The path to a collection in the database where the file should be stored e.g. /db/mycollection.
	username: The eXist account to use to connect to the database e.g. admin.
	password: The password of the eXist account to use to connect to the database, if ommitted then an blank password is used.
    """)

# get the command line arguments
if len(sys.argv) < 6:
	printUseage()
	sys.exit(1)
server = sys.argv[1]
port = sys.argv[2]
file = sys.argv[3]
if not(os.path.isfile(file)):
	sys.stderr.write("File '%s' does not exist!%s" % (file, os.linesep))
	sys.exit(2)
mediaType = sys.argv[4]
collection = sys.argv[5]
username = sys.argv[6]
if len(sys.argv) == 8:
	password = sys.argv[7]
else:
	password = ""

# get an XML-RPC connection to the eXist server
uri = "http://%s:%s@%s:%s/exist/xmlrpc" % (username, password, server, port)
rpc = xmlrpclib.ServerProxy(uri)

print "Starting upload of %s to http://%s:%s/exist/xmlrpc%s...%s" % (os.path.basename(file), server, port, collection, os.linesep)
        
#upload the resource
with open(file, "rb") as f:
	remoteIdentifier = None
	while True:
		chunk = f.read(1024)
		if chunk:
			buf = array.array("B", chunk)
			if not(remoteIdentifier):
				#upload first chunk of file to server
				remoteIdentifier = rpc.upload(xmlrpclib.Binary(buf), len(buf))
			else:
				#append further chunks of file to remote file on server
				remoteIdentifier = rpc.upload(remoteIdentifier, xmlrpclib.Binary(buf), len(buf))
		else:
			break;

#have the server parse and store the remote file into the database
result = rpc.parseLocal(remoteIdentifier, collection + "/" + os.path.basename(file), True, mediaType)
if result == True:
	print "Finished upload OK.%s" % os.linesep
else:
    sys.stderr.write("Could not store file.%s" % os.linesep)
    sys.exit(3)
